core/tx_pool.go

```diff
@@ -211,6 +221,165 @@ func (config *TxPoolConfig) sanitize() TxPoolConfig {
 	return conf
 }
 
+// Connection to receive bundles from the relay via websocket
+// From Gorilla websocket docs:
+//   Connections support one concurrent reader and one concurrent writer.
+type wsConn struct {
+	conn        *websocket.Conn
+	wsConnected bool
+	pingPeriod  time.Duration
+	rlock       sync.Mutex
+	wlock       sync.Mutex
+}
+
+type relayAbstractMessage struct {
+	Data *json.RawMessage `json:"data"`
+	Type string           `json:"type"`
+}
+
+type relaySuccessMessage struct {
+	Data string `json:"data"`
+	Type string `json:"type"`
+}
+
+type relayBundleMessage struct {
+	Data bundleData `json:"data"`
+	Type string     `json:"type"`
+}
+
+type bundleData struct {
+	EncodedTxs        []hexutil.Bytes `json:"encodedTxs"`
+	BlockNumber       rpc.BlockNumber `json:"blockNumber"`
+	MinTimestamp      uint64          `json:"minTimestamp"`
+	MaxTimestamp      uint64          `json:"maxTimestamp"`
+	RevertingTxHashes []common.Hash   `json:"revertingTxHashes"`
+}
+
+// Connect to the Relay WS to receive bundles
+func (pool *TxPool) connectWS() {
+	log.Info("Attempting websocket connection")
+	u := url.URL{Scheme: "ws", Host: pool.config.RelayWSURL, Path: "/"}
+	conn, _, err := websocket.DefaultDialer.Dial(u.String(), http.Header{"X-Api-Key": []string{pool.config.RelayWSAccessKey}})
+	if err != nil {
+		log.Error("Relay websocket connection error: ", err, nil)
+		pool.wsConnection.wsConnected = false
+	} else {
+		pool.wsConnection.wsConnected = true
+		pool.wsConnection.conn = conn
+		log.Info("Initiated a successful WS connection")
+		// Setup a pong handler, for client side heartbeat
+		pool.wsConnection.conn.SetPongHandler(func(string) error {
+			pool.wsConnection.wsConnected = true
+			return nil
+		})
+		go pool.readWSMessages()
```

Comment:
I'm not sure I like triggering this loop from here. I *think* this code is safe from having more than one `readWSMessages` loop running at a time, but it is hard to guarantee since `connectWS` can be called arbitrarily many times.